### ISSUE TITLE
Pause Mechanism for Smart Contract Functions

### ISSUE THREAD

**Issue by njrapidinnovation** [OpenZeppelin/openzeppelin-contracts#4120]
**🧐 Motivation**
I am making a smart contract that has over 100 functions, some of which are dependent on each other. However, since the current open-zeppelin contracts only support full contract pausing and unpausing, I am unable to implement a pause mechanism for specific functions.

I require the ability to pause a particular function, regardless of its visibility.

For example:

```
function test1(){
  test2()
}

function test2(){
} 
```

In the above function, if `test2` is paused and `test1` is called, it should `throw an error indicating that test2` is currently paused.

**Comment by frangio** [comment]
Thanks for the suggestion. This is something we want to look into.

**Comment by njrapidinnovation** [comment]
Thanks @frangio . I have created a solution using `msg.sig` but it fails during internal function calls. Is there any way to fetch the signature of the function which we are calling during internal call in the contract?

**Comment by frangio** [comment]
No, that's not possible. To pause internal functions selectively you may need to add some sort of identifier manually in the modifier.

**Comment by njrapidinnovation** [comment]
Yeah I completely agree with you . Is there any way to fetch the signature of called function during internal calls ? I think it is available in stack but not able to find a way to fetch it?

**Comment by frangio** [comment]
Internally functions are not identified by their signature... They are compiled to a location in the code and the contract will simply do a JUMP into that location.

**Comment by therealharpaljadeja** [comment]
wouldn't `test1` revert anyway when it arrives at `test2()` statement?

**Comment by njrapidinnovation** [comment]
@therealharpaljadeja it can if there is specific check for `test2` function signature inside `test1` . 

But in case of many internal calls we have to make specific check for each function call signature which is not a good approach.
